### PR TITLE
dialects: (stim) initialise dialect

### PR DESCRIPTION
**Pull request by kimxworrall**
Initial commit to add a dialect stim to represent Stim.

Stim is a tool for simulating quantum stabilizer circuits, particularly for quantum error correction (QEC) circuits.

[Find the github library for Stim here](url)

### CODE REVIEW

File: xdsl/dialects/stim.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        super().__init__(regions =[body])
+
+Stim = Dialect(
+    "stim",
+    #first list operations to include in the dialect
+    [
+        StimModuleOp,
+
+    ]
+)
```
Comment (alexarice) [inline xdsl/dialects/stim.py:32]: @superlopuh will be angry at you for removing trailing whitespace

File: xdsl/dialects/stim.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+class StimModuleOp(IRDLOperation):
+    """
+    Base operation containing a stim program
+    """
+
+    name = "stim.circuit"
+
+    body = region_def()
+
+    assembly_format = "$body attr-dict"
```
Comment (alexarice) [inline xdsl/dialects/stim.py:20]: I think this should be `attr-dict-with-keyword`

**Review by alexarice — CHANGES_REQUESTED** [review]

File: xdsl/dialects/stim.py (hunk 1 above)
Comment (superlopuh) [reply xdsl/dialects/stim.py:32]: With the new precommit fixing this should just be a matter of `make precommit`

File: xdsl/dialects/stim.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    assembly_format = "$body attr-dict-with-keyword"
+
+    def __init__(self, body: Region):
+        super().__init__(regions=[body])
+
+
+Stim = Dialect(
+    "stim",
+    # first list operations to include in the dialect
```
Comment (superlopuh) [inline xdsl/dialects/stim.py:27]:
```suggestion
```

**Review by superlopuh — APPROVED** [review]
Beautiful

**Comment by superlopuh** [comment]
Oh except that the tests don't pass :)

File: tests/filecheck/dialects/stim/stim_ops.mlir [hunk 4]
```diff
@@ -4,12 +4,17 @@
 // CHECK:       builtin.module {
 // CHECK-GENERIC:       "builtin.module"() ({
 
+stim.circuit {}
+// CHECK-NEXT:    stim.circuit {
+// CHECK-GENERIC-NEXT:    "stim.circuit"() ({
+// CHECK-NEXT: }
+// CHECK-GENERIC-NEXT:   }) : () -> ()
```
Comment (superlopuh) [inline tests/filecheck/dialects/stim/stim_ops.mlir:11]:
```suggestion
stim.circuit {}
// CHECK-NEXT:    stim.circuit {
// CHECK-NEXT: }
// CHECK-GENERIC-NEXT:    "stim.circuit"() ({
// CHECK-GENERIC-NEXT:   }) : () -> ()
```

File: tests/filecheck/dialects/stim/stim_ops.mlir [hunk 5]
```diff
@@ -4,12 +4,17 @@
...
 
-stim.circuit attributes {} {}
-// CHECK-NEXT:    stim.circuit attributes {} {}
-// CHECK-GENERIC-NEXT:    "stim.circuit"() {} {} : () -> ()
-
+stim.circuit attributes {"hello" = "world"} {}
+// CHECK-NEXT:    stim.circuit attributes {"hello" = "world"} {
+// CHECK-GENERIC-NEXT:    "stim.circuit"() ({
+// CHECK-NEXT: }
+// CHECK-GENERIC-NEXT:   }) {"hello" = "world"} : () -> ()
```
Comment (superlopuh) [inline tests/filecheck/dialects/stim/stim_ops.mlir:17]:
```suggestion
// CHECK-NEXT:    stim.circuit attributes {"hello" = "world"} {
// CHECK-NEXT: }
// CHECK-GENERIC-NEXT:    "stim.circuit"() ({
// CHECK-GENERIC-NEXT:   }) {"hello" = "world"} : () -> ()
```

File: xdsl/dialects/stim/ops.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    def verify(self, verify_nested_ops: bool = True) -> None:
+        #if the region is not empty, check that all of the operations have printing
+        if self.body.blocks.__len__() == 1:
+            for op in self.body.ops:
+                if not isinstance(op, StimOp):
+                    raise (
+                        PyRDLOpDefinitionError(
+                            f"All operations in a stim circuit must sublcass StimOp, found {op.name}"
+                        )
+                    )
```
Comment (superlopuh) [inline xdsl/dialects/stim/ops.py]: I would recommend against this, it makes it harder to lower incrementally, I would let things fail during printing or rewriting

File: xdsl/dialects/stim/ops.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        
+
+    def print_stim(self, printer: StimPrinter):
+        if self.body.blocks.__len__() == 1:
+            for op in self.body.block.ops:
+                op.print
+        printer.print_string("")
```
Comment (superlopuh) [inline xdsl/dialects/stim/ops.py:49]:
```suggestion
    def print_stim(self, printer: StimPrinter):
        for op in self.body.block.ops:
            if not isinstance(op, StimPrintable):
                raise ValueError(f"Cannot print in stim format: {op}")
        printer.print_string("")
```
Would be great to have a test for this that checks the `ValueError`

**Review by alexarice — COMMENTED** [review]
Just to check, is this currently blocked on https://github.com/xdslproject/xdsl/pull/3106 ?

File: xdsl/dialects/stim/ops.py (hunk 7 above)
Comment (kimxworrall) [reply xdsl/dialects/stim/ops.py]: Have added one now in test_stim_printer_parser

**Comment by kimxworrall** [comment]
> Just to check, is this currently blocked on #3106 ?

Not to my knowledge

**Comment by superlopuh** [comment]
This suddenly has way more things, was that intentional?

**Review by superlopuh — APPROVED** [review]
Perfect, thank you!

**Review by alexarice — APPROVED** [review]
Are the printable things for printing to the format that stim expects?

**Comment by kimxworrall** [comment]
> Are the printable things for printing to the format that stim expects?

Yes, I'll add a comment about that in the current pr - the next one will be a parser from stim.

**Pull request merged**